README.md
# Update Git Release Tag Ruby Gem
[![Project Release](https://img.shields.io/badge/release-v2.0.0-blue.svg)](https://github.com/bonusbits/update_git_release_tag)
[![Circle CI](https://circleci.com/gh/bonusbits/update_git_release_tag.svg?style=shield)](https://circleci.com/gh/bonusbits/update_git_release_tag)
[![Join the chat at https://gitter.im/bonusbits](https://badges.gitter.im/bonusbits/bonusbits.svg)](https://gitter.im/bonusbits?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![GitHub issues](https://img.shields.io/github/issues/bonusbits/update_git_release_tag.svg)](https://github.com/bonusbits/update_git_release_tag/issues)

## Description
Ruby Gem to quickly renew a git tag to a specific release locally and remotely. Such as, refreshing a tag named *latest* to the latest release.

## Installation

### Option 1 (rubygem.org) 
Install from a git URL source repository. No need to pull down the repo and build the gem!

1. Install from rubygem.org
    1. ```gem install update_git_release_tag --no-rdoc --no-ri```

### Option 2 (Use specific_install gem) 
Install from a git URL source repository. No need to pull down the repo and build the gem!

1. Install '''specific_install''' gem
    1. ```gem install specific_install --no-rdoc --no-ri```
2. Install gem from git source
    1. ```gem specific_install https://github.com/bonusbits/update_git_release_tag.git```

### Option 2 (Bundler/GemFile)
1. Install bundler gem if missing
    1. ```gem install bundler --no-rdoc --no-ri```
2. Add the gem to your GemFile
    1. ```vim path/to/project/GemFile```
    2. ```gem 'GEMNAME', :github => 'bonusbits/update_git_release_tag'```
    <br>**OR**
    1. ```gem 'GEMNAME', :git => 'git://github.com/bonusbits/update_git_release_tag.git'```
3. Run Bundler from the project folder where the GemFile resides.
    1. ```bundle install```

### Option 3 (Rake Task)
1. Clone the repository
    1. ```git clone https://github.com/bonusbits/update_git_release_tag.git```
2. Change directories to the root of the cloned repository    
3. Build Gem from source
    1. ```rake install:local```

### Option 4 (Clone/Build)
1. Clone the repository
    1. ```git clone https://github.com/bonusbits/update_git_release_tag.git```
2. Change directories to the root of the cloned repository    
3. Build Gem from source
    1. ```gem build update_git_release_tag.gemspec```
4. Install the gem
    1. ```gem install -l /path/to/gem/file/update_git_release_tag.gem```

## Usage

```
Usage: ugrt -r master -t latest [OPTIONS]

Options:
    -r, --release FULLNAME           Release Tag Name. Default is master
    -t, --tag                        Tag to renew (Delete/Create). Default is latest
    -p, --path                       Root Path of Project. Default is Current Working Directory
    -l, --log-level                  Log Output Level. Default is Info
    -h, --help                       (Flag) Show this message
    -v, --version                    (Flag) Output Script Version
```

## Development
After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Bash Script (Deprecated)
Optionally the old BASH script version is available [here](bash_version)

## Contributing
Bug reports and pull requests are welcome on GitHub at https://github.com/bonusbits/update_git_release_tag. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.

## License
The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).


CHANGELOG.md
## CHANGE LOG

## 1.2.0 - 10/28/2017 - Levon Becker
* Renamed Bash Script from urgt.sh to ugrt.sh
* Updated Readme
* Added Change Log

## 1.1.0 - 04/28/2017 - Levon Becker
* Refactored BASH script to Ruby Gem so will work on any OS, have automated testing and easier to download once in rubygem.org.

## 1.0.0 - 03/15/2017 - Levon Becker
* Initial Release

- - -


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)
require 'update_git_release_tag'


spec/update_git_release_tag_spec.rb
require 'spec_helper'

describe UpdateGitReleaseTag do
  it 'has a version number' do
    expect(UpdateGitReleaseTag::VERSION).not_to be nil
  end

  it 'does something useful' do
    expect(false).to eq(true)
  end
end


ugrt.sh
#!/usr/bin/env bash

default_newtag="latest"
script_version="v1.2.0"

function usage () {
usagemessage="
Usage: $0 -r [release_tag_name] -n [new_tag_name]

Options:
    -r Release Tag Name               :  (Required) i.e. v1.0.0
    -n New Tag Name to add/replace    :  (Optional) i.e. latest, rc, dev, or beta. (Defaults to 'latest')
"
    echo "$usagemessage";
}

function version_message() {
versionmessage="Update Latest Tag Version: $script_version"
    echo "$versionmessage";
}

# Argument Parser
while getopts "r:n:vh" opts; do
    case $opts in
        r ) release=$OPTARG;;
        n ) newtag=$OPTARG;;
        v ) version_message; exit 0;;
        h ) usage; exit 0;;
    esac
done

# Bail if Missing Tag Name
if [ -z $release ]; then
    usage
    echo "ERROR: Release Tag Name Missing!"
    exit 1
fi
if [ -z $newtag ]; then newtag=$default_newtag; fi

git pull
git tag -d ${newtag}
git push origin :refs/tags/${newtag}
git checkout ${release}
git tag -af ${newtag} -m "$release Release"
git push origin --tags
git checkout master


lib/update_git_release_tag/cli.rb
require 'update_git_release_tag'
require 'fileutils'
require 'git'
require 'logger'
require 'optparse'
require 'pp'

module UpdateGitReleaseTag
  class Cli < OptionParser
    # Defaults
    @options = Hash.new
    @options['release'] = 'master'
    @options['tag'] = 'latest'
    @options['path'] = Dir.pwd
    @options['log_level'] = 'info'

    options_parser = OptionParser.new do |opts|
      opts.banner = 'Usage: ugrt -r master -t latest [OPTIONS]'
      opts.separator ''
      opts.separator 'Options:'
      opts.on('-r', '--release FULLNAME', 'Release Tag Name. Default is master  ') do |opt|
        @options['release'] = opt
      end
      opts.on('-t', '--tag', 'Tag to renew (Delete/Create). Default is latest') do |opt|
        @options['tag'] = opt
      end
      opts.on('-p', '--path', 'Root Path of Project. Default is Current Working Directory') do |opt|
        @options['path'] = opt
      end
      opts.on('-l', '--log-level', 'Log Output Level. Default is Info') do |opt|
        @options['log_level'] = opt
      end
      opts.on('-h', '--help', '(Flag) Show this message') do
        puts opts
        exit 0
      end
      opts.on('-v', '--version', '(Flag) Output Script Version') do
        puts "Update Git Release Tag v#{UpdateGitReleaseTag::VERSION}"
        exit 0
      end
    end
    options_parser.parse(ARGV)

    @logger = Logger.new(STDOUT)
    @logger.level = begin
                      Object.const_get("Logger::#{@options['log_level'].upcase}")
                    rescue
                      Logger::INFO
                    end
    # @logger.level = Logger::DEBUG
    @logger.level =
      case @options['log_level']
      when 'info'
        Logger::INFO
      when 'warn'
        Logger::WARN
      when 'debug'
        Logger::DEBUG
      when 'fatal'
        Logger::FATAL
      else
        Logger::INFO
      end
    @logger.progname = 'git commands'
    @logger.datetime_format = '%Y-%m-%d_%H:%M:%S'
    @logger.formatter = proc do |severity, datetime, progname, msg|
      "[#{datetime}] #{severity}: (#{progname.upcase}) #{msg}\n"
    end
    # e.g. "2005-09-22 08:51:08 +0900: hello world"
    # I, [2017-04-28 17:28:25#23575]  INFO -- git commands: Starting Git

    def self.run
      @logger.info 'Initializing...'
      # @logger.info('git_commands') { 'Initializing...' }
      git = Git.open(@options['path'], log: @logger)
      starting_point = git.branch
      git.pull
      tags = git.tags
      tags.each { |tag| git.delete_tag(@options['tag']) if tag.name.include?('latest') } unless tags.empty?
      git.push('origin', ":refs/tags/#{@options['tag']}", options: [force: true])
      git.checkout(@options['release'])
      git.add_tag(@options['tag'], options: [annotate: true, force: true, message: 'Refreshed Release Tag'])
      git.push('origin', 'master', options: [tags: true, force: true])
      git.checkout(starting_point)
    end
  end
end
